src/providers/log-action.provider.ts

```diff
@@ -0,0 +1,77 @@
+// Copyright IBM Corp. 2017. All Rights Reserved.
+// Node module: loopback4-example-log-extension
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {inject, Provider, Constructor, Getter} from '@loopback/context';
+import {CoreBindings} from '@loopback/core';
+import {OperationArgs, ParsedRequest} from '@loopback/rest';
+import {getLogMetadata} from '../decorators/log.decorator';
+import {EXAMPLE_LOG_BINDINGS, LOG_LEVEL} from '../keys';
+import {LogFn, Time, TimerFn} from '../types';
+
+export class LogActionProvider implements Provider<LogFn> {
+  constructor(
+    @inject.getter(CoreBindings.CONTROLLER_CLASS)
+    private readonly getController: Getter<Constructor<{}>>,
+    @inject.getter(CoreBindings.CONTROLLER_METHOD_NAME)
+    private readonly getMethod: Getter<string>,
+    @inject(EXAMPLE_LOG_BINDINGS.APP_LOG_LEVEL)
+    private readonly logLevel: number,
+    @inject(EXAMPLE_LOG_BINDINGS.TIMER) public timer: TimerFn,
+  ) {}
+
+  value() {
+    const fn = <LogFn>(async (
```

Comment:
I find the necessity to case the function to `LogFn` as rather suspicious. Why is this cast necessary, assuming the anonymous function has the same signature as LogFn? Is it perhaps because the return type of this anonymous function is not specified?

Reply:
The reason for this "hack" of sorts is because of the unique Signature of `LogFn` (I wasn't even aware of this being a possibility) ... but it's a function with it's own property for another function (`startTimer`). Because if this if the type isn't cast then it results in an error when trying to add `startTimer` as follows: 
```
Property 'startTimer' does not exist on type '(req: ParsedRequest, args: any[], result: any, start?: [number, number] | undefined) => Promise<v...'.
```

Reply:
Oh, of course! I forgot about `startTimer` property, you are right.